### PR TITLE
Speed up `TokenStream` concatenation

### PR DESCRIPTION
**Pull request by nnethercote**
This PR fixes the quadratic behaviour identified in #65080.

r? @Mark-Simulacrum

### CODE REVIEW

**Comment by nnethercote** [comment]
In my local measurements, this provided a very small (<1%) improvement on some of the standard benchmarks.

@bors try @rust-timer queue

File: src/libsyntax/tokenstream.rs [hunk 1]
```diff
@@ -249,20 +249,47 @@ impl TokenStream {
...
-                let mut vec = Vec::with_capacity(tree_count);
 
-                for stream in streams {
-                    match stream.0 {
-                        None => {},
-                        Some(stream2) => vec.extend(stream2.iter().cloned()),
+                // Get the first stream. If it's `None`, create an empty
+                // stream.
+                let mut iter = streams.drain();
+                let mut first_stream_lrc = match iter.next().unwrap().0 {
```
Comment (Centril) [inline src/libsyntax/tokenstream.rs]:
```suggestion
                let mut first_stream_lrc = match iter.next().unwrap().0.unwrap_or_default();
```

File: src/libsyntax/tokenstream.rs [hunk 2]
```diff
@@ -249,20 +249,47 @@ impl TokenStream {
...
+                    Some(first_stream_lrc) => first_stream_lrc,
+                    None => Lrc::new(vec![]),
+                };
+
+                // Append the elements to the first stream, after reserving
+                // space for them.
+                let first_vec_mut = Lrc::make_mut(&mut first_stream_lrc);
+                first_vec_mut.reserve(num_appends);
+                for stream in iter {
```
Comment (Centril) [inline src/libsyntax/tokenstream.rs]: I think this would be cleaner using `.filter_map(|x| x)` and `.flat_map(|s| s.iter().cloned())`.

Comment (bluss) [reply src/libsyntax/tokenstream.rs]: Current version is easy to read, and the alternatives seem to spend more Rust functions to do the same thing, IMO one can prefer the current.

**Review by Mark-Simulacrum — APPROVED** [review]
Code changes look good to me. I'd like to get @petrochenkov or perhaps @matklad to sign off too though since I'm not too familiar with this code.

**Comment by matklad** [comment]
LGTM

**Comment by nnethercote** [comment]
The perf run didn't work. Let's try it the old-fashioned way:

@rust-timer build c25542724f393647fc93a8a0319edaa827e701d9

**Comment by mati865** [comment]
@nnethercote it worked but still haven't finished, you can see it here https://perf.rust-lang.org/status.html

That 700000% wall time regression made queue a bit long...

**Comment by nnethercote** [comment]
Lots of small (<1%) improvements, mostly in the short-running benchmarks.

**Pull request merged**